runtime/kphp_tracing.cpp

```diff
@@ -0,0 +1,729 @@
+
+constexpr int SPECIAL_ENUM_FUNC_AGGREGATES = 1;
+constexpr int SPECIAL_ENUM_TL_MAGICS = 65001;
+
+static bool worker_process_has_sent_enums = false;
+
+tracing_binary_buffer cur_binlog;
+int cur_trace_level;
+static int cur_last_span_id;
+static double cur_start_timestamp;
+static tracing_binary_buffer flush_strings_binlog;
+static class_instance<C$KphpDiv> cur_trace;
+static on_trace_finish_callback_t cur_on_finish_callback;
+static on_trace_enums_callback_t cur_on_enums_callback;
+static on_rpc_provide_details_typed_t cur_on_rpc_details_typed_callback;
+static on_rpc_provide_details_untyped_t cur_on_rpc_details_untyped_callback;
+
+int64_t generate_uniq_id() {
+  return f$mt_rand();
+}
+
+void init_tracing_lib() {
+  cur_trace_level = -1;
+  cur_last_span_id = 0;
+  cur_start_timestamp = 0.0;
+  flush_strings_binlog.set_use_heap_memory();
+}
+
+void free_tracing_lib() {
+  cur_binlog.clear(false);
+  cur_on_finish_callback = {};
+  cur_on_enums_callback = {};
+  cur_on_rpc_details_typed_callback = {};
+  cur_on_rpc_details_untyped_callback = {};
+  cur_trace = class_instance<C$KphpDiv>{};
+  free_tracing_binlog_lib();
+}
+
+// fill the same enum as `@kphp-tracing aggregate` codegens -
+// but for built-in functions that are traced
+static void fill_BuiltinFuncID_strings(array<string> &out) noexcept {
+  auto add = [&out](BuiltinFuncID func_id, const char *title) {
+    out.set_value(static_cast<int64_t>(func_id), string(title, strlen(title)));
+  };
+  auto add_branch = [&out](BuiltinFuncID func_id, BuiltinFuncID branch_num, const char *title) {
+    out.set_value(static_cast<int64_t>(func_id) + (static_cast<int64_t>(branch_num) << static_cast<int>(BuiltinFuncID::_shift_for_branch)), string(title, strlen(title)));
+  };
+
+  out.set_value(0, string("unknown", 7));
+
+  add(BuiltinFuncID::preg_match, "preg_match");
+  add_branch(BuiltinFuncID::preg_match, BuiltinFuncID::branch_regex_needs_compilation, "needs compilation");
+  add(BuiltinFuncID::preg_match_all, "preg_match_all");
+  add_branch(BuiltinFuncID::preg_match_all, BuiltinFuncID::branch_regex_needs_compilation, "needs compilation");
+  add(BuiltinFuncID::preg_split, "preg_split");
+  add_branch(BuiltinFuncID::preg_split, BuiltinFuncID::branch_regex_needs_compilation, "needs compilation");
+  add(BuiltinFuncID::preg_replace, "preg_replace");
+  add_branch(BuiltinFuncID::preg_replace, BuiltinFuncID::branch_regex_needs_compilation, "needs compilation");
+  add(BuiltinFuncID::preg_replace_callback, "preg_replace_callback");
+  add_branch(BuiltinFuncID::preg_replace_callback, BuiltinFuncID::branch_regex_needs_compilation, "needs compilation");
+
+  add(BuiltinFuncID::exec, "exec");
+  add(BuiltinFuncID::system, "system");
+  add(BuiltinFuncID::passthru, "passthru");
+
+  add(BuiltinFuncID::aggregate_regexp_functions, "/regexp/");
+}
+
+static inline double calc_now_timestamp() noexcept {
+  return std::chrono::duration<double>{std::chrono::system_clock::now().time_since_epoch()}.count();
```

Comment:
I am not sure about using it. Considering https://en.cppreference.com/w/cpp/chrono/high_resolution_clock, "its usage generally should be avoided". It's either an alias for `system_clock`, or if not, I am not sure that it won't affect performance, because we need clock ticks thousands of times during script execution.

We don't need very precise time, because anyway, it's converted to `float` to be saved (see `calc_time_offset()`)